composer/lib/dependabot/composer/metadata_finder.rb

```diff
@@ -28,27 +28,26 @@ def look_up_source_from_packagist
         return nil if packagist_listing&.fetch("packages", nil) == []
         return nil unless packagist_listing&.dig("packages", dependency.name.downcase)
 
-        version_listings =
-          packagist_listing["packages"][dependency.name.downcase].
-          select { |version, _| Composer::Version.correct?(version) }.
-          sort_by { |version, _| Composer::Version.new(version) }.
-          map { |_, listing| listing }.
-          reverse
-
-        potential_source_urls =
-          version_listings.
-          flat_map { |info| [info["homepage"], info.dig("source", "url")] }.
-          compact
-
-        source_url = potential_source_urls.find { |url| Source.from_url(url) }
-
-        Source.from_url(source_url)
+        version_listings = packagist_listing["packages"][dependency.name.downcase]
+        # Packagist returns an array of version listings sorted newest to oldest.
+        # So iterate until we find the first URL that appears to be a source URL.
+        #
+        # NOTE: Each listing may not have all fields because they are minified to remove duplicate elements:
+        # * https://github.com/composer/composer/blob/main/UPGRADE-2.0.md#for-composer-repository-implementors
+        # * https://github.com/composer/metadata-minifier
+        version_listings.each do |i|
+          [i["homepage"], i.dig("source", "url")].each do |url|
+            source_url = Source.from_url(url)
+            return source_url unless source_url.nil?
+          end
+        end
+        return nil # rubocop:disable Style/RedundantReturn
       end
 
       def packagist_listing
         return @packagist_listing unless @packagist_listing.nil?
 
-        response = Dependabot::RegistryClient.get(url: "https://repo.packagist.org/p/#{dependency.name.downcase}.json")
+        response = Dependabot::RegistryClient.get(url: "https://repo.packagist.org/p2/#{dependency.name.downcase}.json")
```

Comment:
One little catch here, the v2 api does separate tagged releases and branch releases. So for packages which only have dev versions like https://packagist.org/packages/roave/security-advisories for example you will find no version in this file https://repo.packagist.org/p2/roave/security-advisories.json and instead you should look up in `p2/#{dependency.name.downcase}~dev.json` i.e. https://repo.packagist.org/p2/roave/security-advisories~dev.json

You could also first look up the `~dev` file if this is only used to find the source url. It'll typically be a much smaller file as there are often less dev versions than releases, and on packagist.org at least every package should have at least one dev version, but not every package has a release.

Reply:
Hmm... this is a bit of a conundrum... 

1. The `latest_version_finder.rb` code is used when we query packagist for the latest version. So if someone has a `composer.json` that includes `"roave/security-advisories": "dev-latest"`, then when we query for newest version, because we don't query the branch release  `*~dev.json` file, Dependabot will get an empty version list and simply short-circuit. Initially I was thinking this meant we needed to add support for `*~dev.json` files, but on second thought since branch releases by definition don't have "newer versions" that Dependabot can update, probably we don't actually need this?
2. The `metadata_finder.rb` code only gets called to populate the metadata of a PR _after_ we know we have an update. So once again, since dev/branch releases by definition cannot get PR updates, we'll never hit the metadata finder code... so from a correctness standpoint this should be okay?
3. Within `metadata_finder.rb`, there's a small optimization possible to use the branch release file for finding source URL even for normal tagged releases since it's a smaller file to download, but IIUC that's an optimization, not a correctness issue?

@Seldaek Can you confirm my understanding ☝️ is correct?

Reply:
Thanks! 

I'll probably ship this as-is then and do step 3 as a further optimization in a follow-on PR once I confirm this is working as expected.

Reply:
I explored this a bit in a follow-on PR, but I think for now the optimization isn't worth shipping--see my comments on the PR for details:
* https://github.com/dependabot/dependabot-core/pull/6360